File: src/outreach_prediction.R
# Load necessary libraries
library(caret)

source("src/utils.R")

# Function to predict outreach success
predict_outreach <- function(model, new_data) {
  predictions <- predict(model, new_data)
  return(predictions)
}

run_outreach_prediction <- function(
  model_path = "models/outreach_model.rds",
  new_data_path = "data/new_data.csv",
  output_path = "data/predictions.csv"
) {
  model <- readRDS(model_path)
  new_data <- read.csv(new_data_path)

  predictions <- predict_outreach(model, new_data)

  write.csv(predictions, output_path, row.names = FALSE)
  predictions
}

if (sys.nframe() == 0) {
  run_outreach_prediction()
}

File: src_py/outreach_prediction.py
from __future__ import annotations

from pathlib import Path

from src_py.utils import PROJECT_ROOT, ModelSpec, apply_model, read_csv, read_json, write_csv


def predict(model_path: Path, new_data_path: Path, output_path: Path) -> None:
    model = ModelSpec.from_dict(read_json(model_path))
    rows = read_csv(new_data_path)
    predictions = []
    for row in rows:
        prediction = apply_model(model, row)
        predictions.append({"customer_id": row["customer_id"], "prediction": prediction})
    write_csv(output_path, predictions, ["customer_id", "prediction"])


def run() -> None:
    data_dir = PROJECT_ROOT / "data"
    predict(
        PROJECT_ROOT / "models" / "outreach_model.json",
        data_dir / "new_data.csv",
        data_dir / "predictions.csv",
    )


if __name__ == "__main__":
    run()
